### PR TITLE
Move Windows-specific path validation out of FileStream

### PR DESCRIPTION
**Pull request by stephentoub**
FileStream currently has a few path validation checks that are specific to Windows.  This commit moves these checks into the Windows-specific file system code.

@ericstj, the way I did this does change the order of some exceptions: it's now possible for the ArgumentException for InvalidAppendMode to be thrown before the ArgumentException and NotSupportedException for the additional path validation.  This passed all of the unit tests, so I don't know if that's just because we're missing an important test or because the order doesn't really matter.  I also put the code into the multiplexer, as it's what's shared between Win32 and WinRT file systems... if that's not a good place for it, let me know.

### CODE REVIEW

**Comment by ericstj** [comment]
Unit tests don't cover any of the invalid path scenarios yet because I punted on solving the cross platform problem.  Maybe you can add some tests with this change? :)  Reordering exceptions here does violate the contract and could cause app compat problems.  Perhaps you can instead solve this with just factoring these into a partial method that only gets included for windows builds?  That'll also get the code out of the multiplexer which I'd like to keep simple.

**Comment by stephentoub** [comment]
Ok, I added it to the PAL instead.  This approach ok?

**Comment by ericstj** [comment]
This would break on WinRT since you didn't plumb it through the multiplexer.  Plumbing through the multiplexer seems like unnecessary overhead since each decision made by the multiplexer requires us to hit the disk.  Also we have some assumptions in the multiplexer that we are only getting a valid path.  I was thinking instead we just rename FileStream.handle.*.cs to just FileStream.*.cs and stuff it there.  The FileSystem abstraction is more about things that need to differ dynamically whereas this path difference is something that is just a static difference.  

If we did work to try and share the same implementation of FileSystem across OSes and multiplex dynamically on something like OSVersion.PlatformID I'd have a different opinion but that is off the table right now due to the restrictions that Windows store places on API dependencies in app binaries.

**Comment by stephentoub** [comment]
> This would break on WinRT since you didn't plumb it through the multiplexer

Can we fix the build so that this gets built as well?  It's difficult to verify these kinds of things when both local and CI builds with "build" pass successfully.  Regardless, I'll fix that.

> I was thinking instead we just rename FileStream.handle..cs to just FileStream..cs and stuff it there.

Ok.  I'd considerd that and didn't want to do that because it seems like we have no multiple ways of handling the PAL, one through partial files and one through FileSystem.Current.  It'd be great if we could just have one approach.  But I don't feel strongly about it; if that's what you'd prefer, I'll change it.

**Comment by stephentoub** [comment]
Ok, take 3.  @ericstj, look ok?

File: src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs [hunk 1]
```diff
@@ -16,5 +16,10 @@ public FileStream(SafeFileHandle handle, FileAccess access, int bufferSize, bool
         {
             this._innerStream = new UnixFileStream(handle, access, bufferSize, isAsync, this);
         }
+
+        private static void ValidatePath(string fullPath)
```
Comment (ericstj) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: You can avoid this by declaring private partial static void ValidatePath(string fullPath); in FileStream.cs and adding "partial" to the implementation in FileStream.Win32.cs.  That way no call to ValidatePath is added to the IL for unix build.

Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: Fixed.

**Comment by ericstj** [comment]
LGTM

**Pull request merged**